File: tests/test_state_management.py
# -*- coding: utf-8 -*-
"""
test_state_management.py - Unit tests for state management module
Tests thread safety, pagination, and error handling
"""
import pytest
import threading
import time
from state_management import AttachmentState, ChatHistory, handle_error

class TestAttachmentState:
    """Test AttachmentState class."""
    
    def test_set_and_get(self):
        """Test basic set/get operations."""
        state = AttachmentState()
        state.set("test.txt", "content", "preview")
        
        name, content, preview = state.get()
        assert name == "test.txt"
        assert content == "content"
        assert preview == "preview"
    
    def test_has_attachment(self):
        """Test has_attachment check."""
        state = AttachmentState()
        assert state.has_attachment() is False
        
        state.set("file.txt", "data", "prev")
        assert state.has_attachment() is True
        
        state.clear()
        assert state.has_attachment() is False
    
    def test_clear(self):
        """Test clearing attachment."""
        state = AttachmentState()
        state.set("file.txt", "content", "preview")
        
        state.clear()
        name, content, preview = state.get()
        
        assert name is None
        assert content is None
        assert preview is None
    
    def test_thread_safety(self):
        """Test AttachmentState is thread-safe (CRITICAL TEST)."""
        state = AttachmentState()
        results = []
        errors = []
        
        def set_attachment(i):
            """Set attachment."""
            try:
                state.set(f"file{i}.txt", f"content{i}", f"preview{i}")
                time.sleep(0.001)  # Simulate some work
                name, content, preview = state.get()
                results.append((name, content, preview))
            except Exception as e:
                errors.append(e)
        
        # Run 20 threads concurrently
        threads = [threading.Thread(target=set_attachment, args=(i,)) for i in range(20)]
        
        for t in threads:
            t.start()
        
        for t in threads:
            t.join()
        
        # Should have 20 results, no errors, no corruption
        assert len(errors) == 0, f"Thread safety errors: {errors}"
        assert len(results) == 20
        
        # Final state should be valid (one of the attachments)
        name, content, preview = state.get()
        assert name is not None
        assert content is not None


class TestChatHistory:
    """Test ChatHistory class."""
    
    def test_add_message(self):
        """Test adding messages."""
        history = ChatHistory(max_messages=100)
        history.add('user', 'Hello')
        history.add('assistant', 'Hi there!')
        
        assert history.count() == 2
        messages = history.get_all()
        assert messages[0].role == 'user'
        assert messages[0].content == 'Hello'
        assert messages[1].role == 'assistant'
    
    def test_pagination(self):
        """Test auto-trimming old messages (CRITICAL TEST - prevents memory leaks)."""
        history = ChatHistory(max_messages=100)
        
        # Add 150 messages
        for i in range(150):
            history.add('user', f'Message {i}')
        
        # Should only keep last 100
        assert history.count() == 100
        
        # First message should be #50 (trimmed 0-49)
        messages = history.get_all()
        assert messages[0].content == 'Message 50'
        assert messages[-1].content == 'Message 149'
    
    def test_get_recent(self):
        """Test getting recent N messages."""
        history = ChatHistory()
        
        for i in range(20):
            history.add('user', f'Msg {i}')
        
        recent = history.get_recent(5)
        assert len(recent) == 5
        assert recent[-1].content == 'Msg 19'
        assert recent[0].content == 'Msg 15'
    
    def test_clear(self):
        """Test clearing history."""
        history = ChatHistory()
        history.add('user', 'Test')
        history.add('assistant', 'Response')
        
        assert history.count() == 2
        
        history.clear()
        assert history.count() == 0
        assert len(history.get_all()) == 0
    
    def test_message_timestamps(self):
        """Test messages have timestamps."""
        history = ChatHistory()
        history.add('user', 'Hello')
        
        messages = history.get_all()
        assert hasattr(messages[0], 'timestamp')
        assert messages[0].timestamp > 0


class TestErrorHandling:
    """Test centralized error handling."""
    
    def test_handle_error_logging(self, caplog):
        """Test error is logged."""
        error = ValueError("Test error")
        handle_error(error, "Test Context", notify_user=False)
        
        assert "Test Context" in caplog.text
        assert "ValueError" in caplog.text
    
    def test_error_message_mapping(self):
        """Test different error types get appropriate messages."""
        # This would require mocking ui.notify
        # For now, just verify function doesn't crash
        errors = [
            ConnectionError("Connection failed"),
            TimeoutError("Timeout"),
            ValueError("Invalid value"),
            FileNotFoundError("File missing"),
        ]
        
        for error in errors:
            try:
                handle_error(error, "Test", notify_user=False)
            except Exception as e:
                pytest.fail(f"handle_error raised exception: {e}")


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: rust_output/src/test_state_management.rs
/// test_state_management::py - Unit tests for state management module
/// Tests thread safety, pagination, and error handling

use anyhow::{Result, Context};
use crate::state_management::{AttachmentState, ChatHistory, handle_error};
use std::fs::File;
use std::io::{self, Read, Write};

/// Test AttachmentState class.
#[derive(Debug, Clone)]
pub struct TestAttachmentState {
}

impl TestAttachmentState {
    /// Test basic set/get operations.
    pub fn test_set_and_get(&self) -> () {
        // Test basic set/get operations.
        let mut state = AttachmentState();
        state::set("test.txt".to_string(), "content".to_string(), "preview".to_string());
        let (mut name, mut content, mut preview) = state::get();
        assert!(name == "test.txt".to_string());
        assert!(content == "content".to_string());
        assert!(preview == "preview".to_string());
    }
    /// Test has_attachment check.
    pub fn test_has_attachment(&self) -> () {
        // Test has_attachment check.
        let mut state = AttachmentState();
        assert!(state::has_attachment() == false);
        state::set("file.txt".to_string(), "data".to_string(), "prev".to_string());
        assert!(state::has_attachment() == true);
        state::clear();
        assert!(state::has_attachment() == false);
    }
    /// Test clearing attachment.
    pub fn test_clear(&self) -> () {
        // Test clearing attachment.
        let mut state = AttachmentState();
        state::set("file.txt".to_string(), "content".to_string(), "preview".to_string());
        state::clear();
        let (mut name, mut content, mut preview) = state::get();
        assert!(name.is_none());
        assert!(content.is_none());
        assert!(preview.is_none());
    }
    /// Test AttachmentState is thread-safe (CRITICAL TEST).
    pub fn test_thread_safety(&self) -> Result<()> {
        // Test AttachmentState is thread-safe (CRITICAL TEST).
        let mut state = AttachmentState();
        let mut results = vec![];
        let mut errors = vec![];
        let set_attachment = |i| {
            // Set attachment.
            // try:
            {
                state::set(format!("file{}.txt", i), format!("content{}", i), format!("preview{}", i));
                std::thread::sleep(std::time::Duration::from_secs_f64(0.001_f64));
                let (mut name, mut content, mut preview) = state::get();
                results.push((name, content, preview));
            }
            // except Exception as e:
        };
        let mut threads = 0..20.iter().map(|i| std::thread::spawn(|| {})).collect::<Vec<_>>();
        for t in threads.iter() {
            t.start();
        }
        for t in threads.iter() {
            t.join();
        }
        assert!(errors.len() == 0, "Thread safety errors: {}", errors);
        assert!(results.len() == 20);
        let (mut name, mut content, mut preview) = state::get();
        assert!(name.is_some());
        Ok(assert!(content.is_some()))
    }
}

/// Test ChatHistory class.
#[derive(Debug, Clone)]
pub struct TestChatHistory {
}

impl TestChatHistory {
    /// Test adding messages.
    pub fn test_add_message(&self) -> () {
        // Test adding messages.
        let mut history = ChatHistory(/* max_messages= */ 100);
        history.add("user".to_string(), "Hello".to_string());
        history.add("assistant".to_string(), "Hi there!".to_string());
        assert!(history.count() == 2);
        let mut messages = history.get_all();
        assert!(messages[0].role == "user".to_string());
        assert!(messages[0].content == "Hello".to_string());
        assert!(messages[1].role == "assistant".to_string());
    }
    /// Test auto-trimming old messages (CRITICAL TEST - prevents memory leaks).
    pub fn test_pagination(&self) -> () {
        // Test auto-trimming old messages (CRITICAL TEST - prevents memory leaks).
        let mut history = ChatHistory(/* max_messages= */ 100);
        for i in 0..150.iter() {
            history.add("user".to_string(), format!("Message {}", i));
        }
        assert!(history.count() == 100);
        let mut messages = history.get_all();
        assert!(messages[0].content == "Message 50".to_string());
        assert!(messages[-1].content == "Message 149".to_string());
    }
    /// Test getting recent N messages.
    pub fn test_get_recent(&self) -> () {
        // Test getting recent N messages.
        let mut history = ChatHistory();
        for i in 0..20.iter() {
            history.add("user".to_string(), format!("Msg {}", i));
        }
        let mut recent = history.get_recent(5);
        assert!(recent.len() == 5);
        assert!(recent[-1].content == "Msg 19".to_string());
        assert!(recent[0].content == "Msg 15".to_string());
    }
    /// Test clearing history.
    pub fn test_clear(&self) -> () {
        // Test clearing history.
        let mut history = ChatHistory();
        history.add("user".to_string(), "Test".to_string());
        history.add("assistant".to_string(), "Response".to_string());
        assert!(history.count() == 2);
        history.clear();
        assert!(history.count() == 0);
        assert!(history.get_all().len() == 0);
    }
    /// Test messages have timestamps.
    pub fn test_message_timestamps(&self) -> () {
        // Test messages have timestamps.
        let mut history = ChatHistory();
        history.add("user".to_string(), "Hello".to_string());
        let mut messages = history.get_all();
        assert!(/* hasattr(messages[0], "timestamp".to_string()) */ true);
        assert!(messages[0].timestamp > 0);
    }
}

/// Test centralized error handling.
#[derive(Debug, Clone)]
pub struct TestErrorHandling {
}

impl TestErrorHandling {
    /// Test error is logged.
    pub fn test_handle_error_logging(&self, caplog: String) -> () {
        // Test error is logged.
        let mut error = ValueError("Test error".to_string());
        handle_error(error, "Test Context".to_string(), /* notify_user= */ false);
        assert!(caplog.text.contains(&"Test Context".to_string()));
        assert!(caplog.text.contains(&"ValueError".to_string()));
    }
    /// Test different error types get appropriate messages.
    pub fn test_error_message_mapping(&self) -> Result<()> {
        // Test different error types get appropriate messages.
        let mut errors = vec![ConnectionError("Connection failed".to_string()), TimeoutError("Timeout".to_string()), ValueError("Invalid value".to_string()), FileNotFoundError("File missing".to_string())];
        for error in errors.iter() {
            // try:
            {
                handle_error(error, "Test".to_string(), /* notify_user= */ false);
            }
            // except Exception as e:
        }
    }
}
